ruby/src/main/java/com/google/protobuf/jruby/RubyMap.java

```diff
@@ -189,8 +190,25 @@ public IRubyObject index(ThreadContext context, IRubyObject key) {
      */
     @JRubyMethod(name = "==")
     public IRubyObject eq(ThreadContext context, IRubyObject _other) {
-        if (_other instanceof RubyHash)
-            return toHash(context).op_equal(context, _other);
+        if (_other instanceof RubyHash){
+            // quick check to see if they are equivalent
+            IRubyObject matched = toHash(context).op_equal(context, _other);
+            if (matched == context.runtime.getTrue()) {
+                return context.runtime.getTrue();
+            }
+
+            RubyHash other = (RubyHash)_other;
+            // we have complex objects, such as a message
+            // lets check to see if they are equivalent
+            for (IRubyObject key : table.keySet()) {
```

Comment:
If we do need this deep comparison -- I think it's missing a check for the case where `other` has additional keys. E.g., say that `this` is `{}`, and `other` is `{ 'a' => MyMessage.new(...)}` -- the above `op_equal` returns false, so we fall through to here, the for-loop never iterates (`table` is empty), and we return `true`.